### PR TITLE
feat: allowing routes to be filtered and adding basic example

### PR DESCRIPTION
**Pull request by erickzanardo**
## Status

**READY**

## Description

This PR adds a complete basic authentication example.

While developing the example, I noticed that it was impossible to make a resource that had authenticated methods, while also having unauthenticated ones, which makes sense thinking in a `users` resource, where the `POST /users`, which would be the "register" endpoint, needs to allow unauthenticated calls as an user does not exists yet.

So this PR also bundles a proposal for such feature.

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: .github/workflows/dart_frog_auth.yaml
Comment (alestiago) [inline .github/workflows/dart_frog_auth.yaml:30]: I didn't know there was something wrong with the action. Should this change be in a different PR? Either way, we don't need to comment it out we can simply add a condition that is always false to skip it, [here](https://github.com/VeryGoodOpenSource/very_good_flutter_plugin/blob/20e47783fd7d94259179c17fa56f869c1d9e696c/.github/workflows/my_plugin.yaml#L38) is an example.

```suggestion
  pana:
    uses: VeryGoodOpenSource/very_good_workflows/.github/workflows/pana.yml@v1
    with:
      if: false # TODO(erickzanardo): temporarirly disabled while we investigate what is wrong with the pana action.
      working_directory: packages/dart_frog_auth
```

File: examples/basic_authentication/README.md
Comment (scarletteliza) [inline examples/basic_authentication/README.md:7]:
```suggestion
An example on how to use basic authentication with Dart Frog.
```

File: packages/dart_frog_auth/README.md
Comment (scarletteliza) [inline packages/dart_frog_auth/README.md:152]:
```suggestion
### Filtering Routes
```

File: packages/dart_frog_auth/README.md
Comment (scarletteliza) [inline packages/dart_frog_auth/README.md:158]:
```suggestion
One of those can be described by looking at implementing a basic RESTful CRUD API. In order to make
such an API that allows consumers to create, update, delete, and get user information, the following list
of routes will need to be created:
```

File: packages/dart_frog_auth/README.md
Comment (scarletteliza) [inline packages/dart_frog_auth/README.md:163]:
```suggestion
 - `POST /users`: Creates a user
 - `PATCH /users/[id]`: Updates the user with the given id.
 - `DELETE /users/[id]`: Deletes the user with the given id.
 - `GET /users/[id]`: Returns the user with the given id.
```

File: packages/dart_frog_auth/README.md
Comment (scarletteliza) [inline packages/dart_frog_auth/README.md:165]:
```suggestion
Those endpoints can be translated to the following structure in a Dart Frog backend:
```

File: packages/dart_frog_auth/README.md
Comment (scarletteliza) [inline packages/dart_frog_auth/README.md:218]:
```suggestion
`dart_frog_auth` only solves the authentication part of the problem. To enforce
authorization, it is up to the developer to implement it manually, or use an authorization issue
system like OATH, for example.

In technical terms, a request should return `401` (Unauthorized) when authentication fails and
`403` (Forbidden) when authorization failed.

The following snippet shows how authorization can be manually checked in `DELETE /users/[id]` route,
where only the current logged user is allowed to delete itself:
```

File: packages/dart_frog_auth/README.md
Comment (scarletteliza) [inline packages/dart_frog_auth/README.md:150]: I have a few changes I'd like to make to this README so I'll wait until after this is merged to do so

File: examples/basic_authentication/lib/user_repository.dart
Comment (alestiago) [inline examples/basic_authentication/lib/user_repository.dart:7]:
```suggestion
final List<User> _database = [];
```

I would avoid the abbreviation to avoid confusion. Something similar is also mentioned on [Effective Dart](https://dart.dev/effective-dart/documentation#avoid-abbreviations-and-acronyms-unless-they-are-obvious).

File: examples/basic_authentication/lib/user_repository.dart
Comment (alestiago) [inline examples/basic_authentication/lib/user_repository.dart:21]: I think it makes more sense to just make `_db` visible for testing.

File: examples/basic_authentication/lib/user_repository.dart
Comment (alestiago) [inline examples/basic_authentication/lib/user_repository.dart:44]:
```suggestion
  /// The user's password, in a hashed form.
```

File: examples/basic_authentication/lib/user_repository.dart
Comment (alestiago) [inline examples/basic_authentication/lib/user_repository.dart:68]: This statement would become more efficient and readable if we consider `_db` as a Map (or HashMap precisely).

File: examples/basic_authentication/pubspec.yaml
Comment (alestiago) [inline examples/basic_authentication/pubspec.yaml:2]:
```suggestion
description: An example of dart_frog_auth using the bearer authentication scheme.
```

File: examples/basic_authentication/pubspec.yaml
Comment (alestiago) [inline examples/basic_authentication/pubspec.yaml:1]: I think we also need to update the name. I checked the `hello_world` example and it does have an altered named, see [here](https://github.com/VeryGoodOpenSource/dart_frog/blob/d5ab3fc4ef86bf06021cfb0d443169109574f0d2/examples/hello_world/pubspec.yaml#L1).

```suggestion
name: basic_authentication
```

File: examples/basic_authentication/pubspec.yaml
Comment (alestiago) [inline examples/basic_authentication/pubspec.yaml:19]: Can we use very good analysis v5?

File: examples/basic_authentication/routes/users/[id].dart
Comment (alestiago) [inline examples/basic_authentication/routes/users/[id].dart:11]: I love the new pattern matching 💙

File: examples/basic_authentication/routes/users/[id].dart
Comment (alestiago) [inline examples/basic_authentication/routes/users/[id].dart:19]: You, as a user, can only get yourself? Should we add some authorization here? Maybe admin users can retrieve all other users?

File: examples/basic_authentication/routes/users/[id].dart
Comment (alestiago) [inline examples/basic_authentication/routes/users/[id].dart:29]: Am I missing something or is there no authentication needed to update a user?

File: examples/basic_authentication/routes/users/_middleware.dart
Comment (alestiago) [inline examples/basic_authentication/routes/users/_middleware.dart:21]: I feel like `applyToRoute` is not an ideal name. It is not being applied to a route but to a `Request`. Maybe, we can rename this to simply `applies` or `applyTo`?

File: examples/basic_authentication/test/lib/user_repostiroy_test.dart
Comment (alestiago) [inline examples/basic_authentication/test/lib/user_repostiroy_test.dart:1]: There is a typo in the file name.

File: examples/basic_authentication/test/lib/user_repostiroy_test.dart
Comment (alestiago) [inline examples/basic_authentication/test/lib/user_repostiroy_test.dart:18]: Should we group `userFromCredentials`?

File: examples/basic_authentication/test/lib/user_repostiroy_test.dart
Comment (alestiago) [inline examples/basic_authentication/test/lib/user_repostiroy_test.dart:106]: Shall we group `updateUser`?

Comment (alestiago) [inline examples/basic_authentication/test/lib/user_repostiroy_test.dart:106]:
```suggestion
    test('updateUser updates the user on the database', () async {
```

This is related to one of the previous comments that suggested renaming `db` to `database`. Since, I'm doing an entire review I can't link the comment. This is just a reminder so we don't forger to update the test name abbreviations to.

File: packages/dart_frog_auth/lib/src/dart_frog_auth.dart
Comment (alestiago) [inline packages/dart_frog_auth/lib/src/dart_frog_auth.dart:51]:
```suggestion
  ApplyToRoute applyTo = _defaultApplyToRequest,
```

or 

```suggestion
  ApplyToRoute appliesTo = _defaultAppliesToRequests,
```

or

```suggestion
  ApplyToRoute appliesToRequests = _defaultAppliesToRequests,
```


I personally prefer the plural ones since they infer it can be applied to more than one type of request.

File: packages/dart_frog_auth/lib/src/dart_frog_auth.dart
Comment (alestiago) [inline packages/dart_frog_auth/lib/src/dart_frog_auth.dart:91]:
```suggestion
/// apply to the route and the call will have no authentication validation.
```

File: packages/dart_frog_auth/test/src/dart_frog_auth_test.dart
Comment (alestiago) [inline packages/dart_frog_auth/test/src/dart_frog_auth_test.dart:112]: Should we have a test for  "applies to routes that match the custom predicate"?

File: examples/basic_authentication/lib/user_repository.dart
Comment (alestiago) [reply examples/basic_authentication/lib/user_repository.dart:11]: See https://github.com/VeryGoodOpenSource/dart_frog/pull/715#discussion_r1239557507 first.

File: examples/basic_authentication/lib/user_repository.dart
Comment (alestiago) [reply examples/basic_authentication/lib/user_repository.dart:12]: Related to https://github.com/VeryGoodOpenSource/dart_frog/pull/715#discussion_r1239562535.

File: examples/basic_authentication/lib/user_repository.dart
Comment (alestiago) [reply examples/basic_authentication/lib/user_repository.dart:68]: Related to https://github.com/VeryGoodOpenSource/dart_frog/pull/715#discussion_r1239558068.

File: examples/basic_authentication/lib/user_repository.dart
Comment (alestiago) [reply examples/basic_authentication/lib/user_repository.dart:7]: Related to https://github.com/VeryGoodOpenSource/dart_frog/pull/715#discussion_r1239565037.

File: packages/dart_frog_auth/lib/src/dart_frog_auth.dart
Comment (alestiago) [reply packages/dart_frog_auth/lib/src/dart_frog_auth.dart:51]: Related to https://github.com/VeryGoodOpenSource/dart_frog/pull/715#discussion_r1239582940.

File: examples/basic_authentication/routes/users/_middleware.dart
Comment (alestiago) [reply examples/basic_authentication/routes/users/_middleware.dart:21]: Related to https://github.com/VeryGoodOpenSource/dart_frog/pull/715#discussion_r1239589315.

File: examples/basic_authentication/lib/user_repository.dart
Comment (erickzanardo) [reply examples/basic_authentication/lib/user_repository.dart:7]: Does it matter in this case? This is just a fictional database 😉

Comment (erickzanardo) [reply examples/basic_authentication/lib/user_repository.dart:7]: I also try to avoid abbreviations everytime that is possible, but in this case, I will argue that this is still compliant to effective dart, as we see there:

> AVOID abbreviations and acronyms unless they are obvious

And `db` is super common and pretty obvious term for database abbreviations.

File: examples/basic_authentication/lib/user_repository.dart
Comment (erickzanardo) [reply examples/basic_authentication/lib/user_repository.dart:21]: Why? Doesn't that accomplish the same thing?

File: examples/basic_authentication/lib/user_repository.dart
Comment (erickzanardo) [reply examples/basic_authentication/lib/user_repository.dart:68]: Again, this is a fictional database, perfomance is not a concern here.

File: examples/basic_authentication/routes/users/[id].dart
Comment (erickzanardo) [reply examples/basic_authentication/routes/users/[id].dart:19]: > Should we add some authorization here?

This is authorization code, in this example, only you can get your information, this was a design choice for the example to showcase the difference of authentication vs authorization, which I explain on the README.

> Maybe admin users can retrieve all other users?

Definitely, but I don't want to make this example unnecessarily complicated, this is an example for authentication, not authorization, and even so authorization was already showcased on the GET/PUT/POST /users as a bonus. An admin functionality could be a cool thing to be mentioned in a more complete tutorial thought.

File: examples/basic_authentication/routes/users/[id].dart
Comment (erickzanardo) [reply examples/basic_authentication/routes/users/[id].dart:29]: Not sure I got it, are you asking if the user needs to be authenticated to be updated?

If so, it does and that is already handled by the middleware.

But your comment made me realized that I forgot to check if the authenticated user is the same from the id, so I am gonna address that.

File: examples/basic_authentication/routes/users/_middleware.dart
Comment (erickzanardo) [reply examples/basic_authentication/routes/users/_middleware.dart:21]: `applies` indeed feels like a better name, I am gonna change.

File: examples/basic_authentication/test/lib/user_repostiroy_test.dart
Comment (erickzanardo) [reply examples/basic_authentication/test/lib/user_repostiroy_test.dart:106]: I still think that this is a case where the abbreviation is pretty obvious.

File: packages/dart_frog_auth/test/src/dart_frog_auth_test.dart
Comment (erickzanardo) [reply packages/dart_frog_auth/test/src/dart_frog_auth_test.dart:112]: I don't think we need, since that is the default right? Wouldn't that be a redundant test?

**Review by renancaraujo — APPROVED** [review]

File: examples/basic_authentication/lib/user_repository.dart
Comment (alestiago) [reply examples/basic_authentication/lib/user_repository.dart:21]: The idea behind it is allow moving the test related methods to the test instead of letting them be in the source code. For example, I would expect a method named `mockUserData` to be inside a test file.

File: examples/basic_authentication/routes/users/[id].dart
Comment (alestiago) [reply examples/basic_authentication/routes/users/[id].dart:29]: > But your comment made me realized that I forgot to check if the authenticated user is the same from the id, so I am gonna address that.

Awesome, that's what I was trying to point out but wasn't completely sure.

File: examples/basic_authentication/lib/user_repository.dart
Comment (alestiago) [reply examples/basic_authentication/lib/user_repository.dart:7]: My brain goes through an additional process when reading "db". I read "db" and then translate it to "database" and then continue reading. Whereas, my brain does not do the same when reading "CLI".

I agree "db" is common and hence this would also fit within the Effective Dart guideline. So the change is just personal preference. Feel free to resolve this Nit.

File: examples/basic_authentication/lib/user_repository.dart
Comment (alestiago) [reply examples/basic_authentication/lib/user_repository.dart:11]: Note that, regardless of the https://github.com/VeryGoodOpenSource/dart_frog/pull/715#discussion_r1239557507 comment. If we decide not to move the method to the tests we should move the visibleForTesting tag below the documentation.

File: examples/basic_authentication/lib/user_repository.dart
Comment (alestiago) [reply examples/basic_authentication/lib/user_repository.dart:7]: That's true as you point out the performance in this fictional database is not of a massive worry, since it is an example.   I should have been more clear with the comment. Since my intention pointing out that fact wasn't precisely improving the performance of the example, but leading by example.  Let me rephrase:

I believe the objective of the example is to show users how the tool works without confusion and also take the opportunity to lead and inculcate good practices and behaviours.

Firstly, I think that changing the List to a Map will avoid readers feeling confused when reading the example, here's why:

Throughout the code you'll find that an `id` variable is being assigned when creating users, see [here](https://github.com/VeryGoodOpenSource/dart_frog/blob/f674f3e314b06f416373aa7e5475c3704c0b8f96/examples/basic_authentication/routes/users/index.dart#L22). Developers usually expect "id" to refer to "identifier" and usually identifiers are unique. This leads to the developer thinking that users are unique within the example. However, unintuitively in this example, you can have multiple instances of the same user (with the same name and id) with multiple different passwords! Personally, I think this will confuse some of our readers and using a `Map`-like data structure will solve this strange behaviour.

Secondly, I think that we should try to lead by example and show off good practices in our examples to. In this case, the List is not an appropriate data structure since we are only using it to search within it. I think is a nice bonus to inculcate that a Map is a clearly a better choice for the given scope of the example.

File: examples/basic_authentication/lib/user_repository.dart
Comment (alestiago) [reply examples/basic_authentication/lib/user_repository.dart:68]: Shifting conversation to https://github.com/VeryGoodOpenSource/dart_frog/pull/715#discussion_r1243556038.

File: packages/dart_frog_auth/test/src/dart_frog_auth_test.dart
Comment (alestiago) [reply packages/dart_frog_auth/test/src/dart_frog_auth_test.dart:112]: I think it wouldn't hurt. As long as there's some test checking this would be great. If in addition this tests are explicitly testing the functionality even better.

I currently think we are indeed lacking this test. Simply removing the following from [here](https://github.com/VeryGoodOpenSource/dart_frog/blob/f674f3e314b06f416373aa7e5475c3704c0b8f96/packages/dart_frog_auth/lib/src/dart_frog_auth.dart#L54) still manages to pass all the tests!!

```dart
    if (!await applies(context)) {
            return handler(context);
          }
```

File: examples/basic_authentication/test/lib/user_repostiroy_test.dart
Comment (alestiago) [reply examples/basic_authentication/test/lib/user_repostiroy_test.dart]: This was just a reminder comment in the case we went forward with the change. Hence, resolving and moving the conversation to https://github.com/VeryGoodOpenSource/dart_frog/pull/715#discussion_r1239561733.

File: packages/dart_frog_auth/test/src/dart_frog_auth_test.dart
Comment (erickzanardo) [reply packages/dart_frog_auth/test/src/dart_frog_auth_test.dart:112]: Interesting, fair point that the tests fails if we remove those lines, I am gonna check how I can make sure that is not the case.

Comment (erickzanardo) [reply packages/dart_frog_auth/test/src/dart_frog_auth_test.dart:112]: I actually had the test case there already, but it was wrong, I fixed it now makes sure that that code section is being correctly verified, good callout.

File: .github/workflows/dart_frog_auth.yaml
Comment (renancaraujo) [inline .github/workflows/dart_frog_auth.yaml:21]: v1 and v1.13.0 point to the same commit.

File: .github/workflows/examples_basic_authentication.yaml
Comment (renancaraujo) [inline .github/workflows/examples_basic_authentication.yaml:15]:
```suggestion
      - ".github/workflows/examples_basic_authentication.yaml"
```

File: .github/workflows/examples_basic_authentication.yaml
Comment (renancaraujo) [inline .github/workflows/examples_basic_authentication.yaml:19]: This should also be triggered by changes in the authentication package?

File: examples/basic_authentication/pubspec.yaml
Comment (renancaraujo) [inline examples/basic_authentication/pubspec.yaml:10]: bro

**Review by renancaraujo — DISMISSED** [review]

File: examples/basic_authentication/lib/user_repository.dart
Comment (alestiago) [inline examples/basic_authentication/lib/user_repository.dart:9]: Nit: avoid dangling tags
```suggestion
/// In memory database of users.
@visibleForTesting
```

File: examples/basic_authentication/lib/user_repository.dart
Comment (alestiago) [inline examples/basic_authentication/lib/user_repository.dart:37]: Nit: to consider two users to be the same only if they share the same id?

```suggestion
  List<Object?> get props => [id];
```

File: examples/basic_authentication/lib/user_repository.dart
Comment (alestiago) [inline examples/basic_authentication/lib/user_repository.dart:93]:
```suggestion
  Future<void> deleteUser(String id) async {
    db.remove(id);
  }
```

File: examples/basic_authentication/lib/user_repository.dart
Comment (alestiago) [inline examples/basic_authentication/lib/user_repository.dart:119]:
```suggestion
  /// Updates the user with the given [id] with the given [name], [username]
  /// and [password] (in raw format).
  Future<void> updateUser({
    required String id,
    required String? name,
    required String? username,
    required String? password,
  }) async {
    final currentUser = db[id];

    if (currentUser == null) {
      return Future.error(Exception('User not found'));
    }

    final user = User(
      id: id,
      name: name ?? currentUser.name,
      username: username ?? currentUser.username,
      password: password != null ? _hashValue(password) : currentUser.password,
    );

    db[id] = user;
  }
```

File: packages/dart_frog_auth/lib/src/dart_frog_auth.dart
Comment (alestiago) [inline packages/dart_frog_auth/lib/src/dart_frog_auth.dart:23]: Since we rename the parameter to `applyToRoute` to `applies` due maybe we should consider renaming the typedef since it is also public?

**Review by alestiago — APPROVED** [review]

File: examples/basic_authentication/lib/user_repository.dart
Comment (erickzanardo) [reply examples/basic_authentication/lib/user_repository.dart:36]: In theory, yes, but we also need to take in account all of the fields, because imagine in a `updateUser` method, we will change some fields of the model given the new input data from the user. Then the id will be the same, but those instances are not equal anymore, even though both instances point to the same user.

File: examples/basic_authentication/pubspec.yaml
Comment (erickzanardo) [reply examples/basic_authentication/pubspec.yaml:10]: ?

File: .github/workflows/dart_frog_auth.yaml
Comment (erickzanardo) [reply .github/workflows/dart_frog_auth.yaml:21]:
```suggestion
    uses: VeryGoodOpenSource/very_good_workflows/.github/workflows/dart_package.yml@v1
```

File: .github/workflows/dart_frog_auth.yaml
Comment (wolfenrain) [inline .github/workflows/dart_frog_auth.yaml:25]:
```suggestion
    uses: VeryGoodOpenSource/very_good_workflows/.github/workflows/pana.yml@v1
```

**Review by wolfenrain — APPROVED** [review]

File: examples/basic_authentication/pubspec.yaml
Comment (renancaraujo) [reply examples/basic_authentication/pubspec.yaml:10]: Joke attempt failed.

**Review by renancaraujo — APPROVED** [review]

**Pull request merged**
